wallet: Add separate balance info for non-mempool wallet txs

ajtowns opened the pull request:
Changes `getbalances` to report the sum of txos spent by transactions that aren't confirmed nor in the mempool (eg due to being part of too long a mempool chain, or spending non-standard outputs, or having a datacarrier output that exceeds `-datacarriersize`, etc). Those values are added to the trusted/untrusted_pending/immature/used fields as appropriate (where previously they were skipped), and subtracted from the new nonmempool field, so that the sum of all fields remains the same.

For example:

```
$ bitcoin-cli -regtest getbalances
{
  "mine": {
    "trusted": 6049.99999220,
    "untrusted_pending": 0.00000000,
    "immature": 3200.00000780,
    "nonmempool": -100.00000000
  },
  "lastprocessedblock": {
    "hash": "3ab4582226d5e8ad76438db48d76e822c31bce2cdbc7ba82a5d974a277515d0d",
    "height": 221
  }
}
```

Closes #11887

vasild reviewed (commented):
Concept ACK

This extends the `getbalances` RPC. Would be good to also update the GUI (not necessary in this PR).

The new type of transactions would have somehow to be fed to the wallet so that they would be returned by `wallet.GetTXOs()`, called by `GetBalance()`. How would that be done?

ajtowns committed:
tests: Add test for mempool-invalid wallet tx

Uses send rpc to create a tx with oversized OP_RETURN output, verifies
that it doesn't enter the mempool, and that getbalance rpc returns a
nonmempool value.

achow101 commented:
> Another approach would be to report it as:
> 
> ```
> trusted: 6049.99
> untrusted_pending: 0
> immature: 3200
> nonmempool-pending: -150
> ```
> 
> ie "your balance is this, but you have X funds that are currently not in your mempool that you might lose some/all of"

I would prefer this approach and docs just need to be clear that some of that amount may come back as change.

ajtowns committed:
wallet: Have GetBalance report used amount directly without two calls

ajtowns committed:
wallet: Add separate balance info for non-mempool wallet txs

achow101 commented on src/wallet/receive.cpp:
```diff
@@ -255,17 +255,25 @@ Balance GetBalance(const CWallet& wallet, const int min_depth, bool avoid_reuse)
+                }
+                if (bucket) {
+                    // Get the amounts for mine
+                    CAmount credit_mine = txo.GetTxOut().nValue;
```
In 81e763f1e5d878383487b5e4e695b3dca703e1d6 "wallet: Have GetBalance report used amount directly without two calls"

This could be brought inline with `*bucket += credit_mine;` below as `credit_mine` is only ever used there.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -738,6 +738,29 @@ bool CWallet::IsSpent(const COutPoint& outpoint) const
+            if (wtx.isConfirmed()) return SpendType::CONFIRMED;
+            if (wtx.InMempool()) {
+                st = SpendType::MEMPOOL;
+            } else if (!wtx.isAbandoned() && !wtx.isBlockConflicted() && !wtx.isMempoolConflicted()) {
```
In 57c9d938824e833fea4f2c2f58ba2800444c23cb "wallet: Add separate balance info for non-mempool wallet txs"

This could be simplified:
```suggestion
            } else if (wtx.isInactive() && !wtx.isAbandoned()) {
```

achow101 reviewed (commented):
ACK f77bf4cf1a510a9b1b50dcbc6d045e28876220fa

whitslack commented:
@achow101: Thank you for responding. I'm replying below.

> > The spending transaction does not appear in `listtransactions`
> 
> It should, but you may need to set a higher `count` parameter.

Nope. `listtransactions` returns the three previous transactions that each created one output to my address, but it does not show the still-unconfirmed transaction that spends those three outputs.

To be clear: I created the spending transaction manually using `bitcoin-tx -create`, deducted a suitable fee from it using `fundrawtransaction` with `{"add_inputs": false, "subtractFeeFromOutputs": [0], "fee_rate": 0.1}`, signed it using `signrawtransactionwithwallet`, and broadcast it with `sendrawtransaction`.

I don't know if this is relevant, but when I called `sendrawtransaction`, I did not specify the wallet to `bitcoin-cli` with any `-rpcwallet=` argument; I just let the RPC go to the default endpoint, assuming that broadcasting a raw transaction is not an operation that requires specifying a wallet.

> To be clear, the wallet never deletes a transaction unless the user has explicitly called `removeprunedfunds`.

That's great. So, how do I query the wallet for a list of the transactions that it knows about but that it's not including in `listunspent`, `listtransactions`, or `getbalances`?

w0xlt reviewed (commented):
lgtm ACK 32325d17777fcb58f36842fc8e679fbedf6b2b02

musaHaruna reviewed (commented):
Code Review ACK [32325d1](https://github.com/bitcoin/bitcoin/pull/33671/changes/32325d17777fcb58f36842fc8e679fbedf6b2b02)

Extends the `getbalances` RPC to account for transactions that are neither in the mempool nor confirmed in a block, by introducing a new nonmempool field.

The `GetBalance()` function works by scanning every UTXO in the wallet and classifying each one into balance buckets based on its state. These buckets represent whether the coin is confirmed (trusted), unconfirmed but present in the mempool (untrusted_pending), or still immature (coinbase outputs). Each UTXO is evaluated independently using its confirmation depth, trust status, and spending status to determine the appropriate category.

The key addition is support for UTXOs that are spent by transactions known to the wallet but neither confirmed nor present in the mempool. These are classified as NONMEMPOOL by `HowSpent()`. In this case, the UTXO is still included in the normal balance classification, but its value is also subtracted from a special adjustment field, `m_mine_nonmempool`. This adjustment ensures that, even in these edge cases where wallet state and mempool/blockchain state diverge, the total sum across all balance fields remains consistent and does not overstate available funds.

Not sure whether the `getbalances` RPC would benefit from additional release notes explaining the new `nonmempool` field.

> Bumped to remove the parts about locked coins. I plan to leave any further simplifications for follow-ups—let me know if that’s not the right approach. 

I’m happy to add those simplifications in a follow-up once this PR is merged.

achow101 commented:
ACK 32325d17777fcb58f36842fc8e679fbedf6b2b02
